rclcpp_action/test/test_server.cpp

```diff
@@ -923,3 +982,259 @@ TEST_F(TestServer, deferred_execution)
+
+  void SetUp()
+  {
+    node_ = std::make_shared<rclcpp::Node>("is_ready_error", "/rclcpp_action/construct");
+    using GoalHandle = rclcpp_action::ServerGoalHandle<Fibonacci>;
+    action_server_ = rclcpp_action::create_server<Fibonacci>(
+      node_, "fibonacci",
+      [](const GoalUUID &, std::shared_ptr<const Fibonacci::Goal>) {
+        return rclcpp_action::GoalResponse::ACCEPT_AND_DEFER;
+      },
+      [](std::shared_ptr<GoalHandle>) {
+        return rclcpp_action::CancelResponse::REJECT;
+      },
+      [this](std::shared_ptr<GoalHandle> handle) {
+        current_goal_handle_ = handle;
+      });
+  }
+
+protected:
+  std::shared_ptr<rclcpp::Node> node_;
+  std::shared_ptr<rclcpp_action::Server<Fibonacci>> action_server_;
+  std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> current_goal_handle_;
+};
+
+TEST_F(TestValidServer, is_ready_rcl_error) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  auto rcl_context = node_->get_node_base_interface()->get_context()->get_rcl_context().get();
+  ASSERT_EQ(
+    RCL_RET_OK,
+    rcl_wait_set_init(&wait_set, 10, 10, 10, 10, 10, 10, rcl_context, allocator));
+  RCLCPP_SCOPE_EXIT(
+  {
+    EXPECT_EQ(RCL_RET_OK, rcl_wait_set_fini(&wait_set));
+  });
+  ASSERT_TRUE(action_server_->add_to_wait_set(&wait_set));
+
+  EXPECT_TRUE(action_server_->is_ready(&wait_set));
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp_action", rcl_action_server_wait_set_get_entities_ready, RCL_RET_ERROR);
+  EXPECT_THROW(action_server_->is_ready(&wait_set), rclcpp::exceptions::RCLError);
+}
+
+class TestBasicServer : public TestServer
+{
+public:
+  void SetUp()
+  {
+    node_ = std::make_shared<rclcpp::Node>("goal_request", "/rclcpp_action/goal_request");
+    uuid_ = {{1, 2, 3, 4, 5, 6, 70, 8, 9, 1, 11, 120, 13, 140, 15, 160}};
+    action_server_ = rclcpp_action::create_server<Fibonacci>(
+      node_, "fibonacci",
+      [](const GoalUUID &, std::shared_ptr<const Fibonacci::Goal>) {
+        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+      },
+      [](std::shared_ptr<GoalHandle>) {
+        return rclcpp_action::CancelResponse::REJECT;
+      },
+      [this](std::shared_ptr<GoalHandle> handle) {
+        goal_handle_ = handle;
+      });
+  }
+
+  virtual void SendClientRequest(
+    std::chrono::milliseconds timeout = std::chrono::milliseconds(-1)) = 0;
+
+protected:
+  GoalUUID uuid_;
+  std::shared_ptr<rclcpp::Node> node_;
+  std::shared_ptr<rclcpp_action::Server<Fibonacci>> action_server_;
+
+  using GoalHandle = rclcpp_action::ServerGoalHandle<Fibonacci>;
+  std::shared_ptr<GoalHandle> goal_handle_;
+};
+
+class TestGoalRequestServer : public TestBasicServer
+{
+public:
+  void SendClientRequest(
+    std::chrono::milliseconds timeout = std::chrono::milliseconds(-1)) override
+  {
+    send_goal_request(node_, uuid_, timeout);
+    goal_handle_->execute();
```

Comment:
@brawner Hi, I found that this throws an exception while writing my test case based on your test. 
If you call `SendClientRequest` without `EXPECT_THROW`, you get 

> C++ exception with description "goal_handle attempted invalid transition from state EXECUTING with event EXECUTE, at /opt/ros/src/ros2/rcl/rcl_action/src/rcl_action/goal_handle.c:95" thrown in the test body.

Without this line, some tests got failed.

```
[1.870s] 2: [  FAILED  ] TestGoalRequestServer.publish_result_send_result_response_errors
[1.870s] 2: [  FAILED  ] TestGoalRequestServer.publish_status_rcl_errors
[1.871s] 2: [  FAILED  ] TestGoalRequestServer.publish_status_send_cancel_response_errors
[1.871s] 2: [  FAILED  ] TestGoalRequestServer.get_result_rcl_errors
[1.871s] 2: [  FAILED  ] TestGoalRequestServer.send_result_rcl_errors
```

Reply:
You're correct. I wrote these tests just to cover different errors that can occur, but it won't work as it is anyway. I think it makes sense for me to add a correctly functioning unit test to prove the test fixture code is good.

Reply:
Ok, I updated these test fixtures. I now run a nominal check that shouldn't throw, so you can refer either `TestGoalRequestServer` or `TestCancelRequestServer` if need be (but it sounds like you were able to write one).